Detectors/EMCAL/simulation/src/RawWriter.cxx

```diff
@@ -244,4 +245,29 @@ std::vector<char> RawWriter::createRCUTrailer(int payloadsize, int feca, int fec
   std::vector<char> encoded(trailerwords.size() * sizeof(uint32_t));
   memcpy(encoded.data(), trailerwords.data(), trailerwords.size() * sizeof(uint32_t));
   return encoded;
+}
+
+
+int RawWriter::carryOverMethod(const header::RDHAny* rdh, const gsl::span<char> data,
+                                            const char* ptr, int maxSize, int splitID,
+                                            std::vector<char>& trailer, std::vector<char>& header) const {
+  int offs = ptr - &data[0]; // offset wrt the head of the payload
```

Comment:
@mfasDa concerning the ``carryOverMethod``: does the original data you provide to ``writer->addData(..)`` already contain the trailer (placeholder) which need to be finalized in  ``carryOverMethod`` or you want to build it on the fly add and to every page payload, regardless if it the data need splitting or not? 
What we discussed yesterday is to call ``carryOverMethod`` also for the last chunk if there was at least 1 splitting, but given this will be wrong if your data comes w/o trailer.
In the latter case, I would simply add an option (you will need to activate it for EMCal) to always call this method, even if splitting is not needed at all.
But if your original data contain a trailer, then it is more tricky, since it will need to be modified in the original data which is supposed to be ``const``...

Additional point: the ``trailer`` string supplied by the method must be a multiple of 16B CRU word.
Is this is a case for the RCU trailer your create?
Also, if your original data do not contain a trailer (so it is added on top of the maxSize): 
can the trailer size exceed one 16B word? In this case, if the ``maxSize`` estimated by the writer is 16B (minimal possible value), then you will get a negative ``actualSize``.

Reply:
@shahor02 Currently the trailer is added at the end of the payload, containing the full payload size which is the number of 32bit ALTRO words in the payload sent to addData. In the hardware the trailer (size typically 9 32 bit ADC words) is added to every page and it contains the number of 10 bit words per page, so in the decoding I have to add the payload size from all the trailers belonging to the same page in order  to know how the exact amount of words. In the simulation the trailers must be handled in the same way.

In the carryOverMethod implementation used now I read the trailer from the end-of-payload and replace the payload size with the payload size of the page, in the same way as it is done for ITSMFT. (This answers the first question with yes) This is rather trivial as the number of 8bit words just needs to be divided by 4. However the last trailer must be adapted as well and this I can only do once I know the size of the payload in the last page.

Concerning maxSize: I was under the impression this is the size of the page? The trailer would always be smaller. Did I misunderstand something?

Reply:
Hi @mfasDa
The maxSize is the number of bytes remaining in the current open page to reach 8KB limit (or the super-page to reach is limit, by default 1MB). So, ultimately, the maxSize can be just 16B. But if your data already contains the fixed-size trailer, this is not a problem. Since you are starting a new page for every trigger, I don't think you will have a situation of only 16B remaining on the 8KB page, it could happen if you add a new payload to a partially filled page. And in case of super-page filling there is a protection to not create a new page in the same super-page if less than 128B left (from which 64 will be taken by the RDH), so, you should be safe.

Basically, for the EMCAL I will need the writer will need to call carryOverMethod for the very last chunk (where no splitting is actually needed) and instead of adding the supplied trailer after writing the payload (as it would do if there is a real splitting) it should overwrite the last ``carryOverTrailer.size()`` bytes, is this correct?

Reply:
@mfasDa please check if https://github.com/AliceO2Group/AliceO2/pull/4179 is what you want. You will need to enable the carry-over call for the last chunk by ``RawFileWriter::setApplyCarryOverToLastPage(true);`` and detect such a call as the modified test shows: https://github.com/AliceO2Group/AliceO2/blob/863c3435580f5fbe1bc34870f4a40a31f45aea82/Detectors/Raw/test/testRawReaderWriter.cxx#L166-L170

Reply:
Hi @shahor02, yes, the conclusion is correct. The carryOverMethod currently sets the payload size to the trailer whenever it is called. So since the end of data contains the trailer either it has to be overwritten by the RawFileWriter or our custom carryOverMethod. In the second case I would not add a trailer but modify it myself (probably overall easier). In this case I would need to know that we are on the last page.

Reply:
@mfasDa as far as I can see, your carryOverMethod sets the returned ``actual`` size to ``maxSize - trailer.size``, this is ok except when you are on the last chunk: in this case the you should return the original ``maxSize``  as an actual payload to write and supply the modified trailer string. The writer will overtwrite last  ``trailer.size()`` of the payload last chunk by this modified trailer. You cannot modify in the carry-over method the original data: it is ``const gsl::span``, just provide an updated trailer to override the one in the payload.

Reply:
OK, so I only need to know that I am on the last page to include the trailer size in the from the actual size. I was thinking to take this from the payload size but in principle the page could be 8 kB. Anything that tells me how much data is still to come? I.e. the gsl::span contains the full data or the data starting from the page?

Reply:
The ``const gsl::span<char> data`` refers to the full original payload you provided to ``addData``, ``const char* ptr``
is start of data yet to write (can be anything between ``data.begin() and data.end()``) and the ``int maxSize`` is the max number of bytes the writer can write (starting from the ``ptr``) to currently open 8kb page.
You can detect that you are on the last page by https://github.com/AliceO2Group/AliceO2/blob/72e4fba376d9cff8b1677b5cd93982123496d9ba/Detectors/Raw/test/testRawReaderWriter.cxx#L166-L167